src/translator/request.cpp

```diff
@@ -11,38 +11,68 @@ namespace marian {
 namespace bergamot {
 
 // -----------------------------------------------------------------
-Request::Request(size_t Id, Segments &&segments, ResponseBuilder &&responseBuilder)
+Request::Request(size_t Id, Segments &&segments, ResponseBuilder &&responseBuilder, TranslationCache *cache)
     : Id_(Id),
       segments_(std::move(segments)),
-      responseBuilder_(std::move(responseBuilder))
+      responseBuilder_(std::move(responseBuilder)),
+      cache_(cache)
 
 {
-  counter_ = segments_.size();
-  histories_.resize(segments_.size(), nullptr);
-
-  // If there are no segments_, we are never able to trigger the responseBuilder
-  // calls from a different thread. However, in this case we want an empty valid
-  // response.
+  // 1. If there are no segments_, we are never able to trigger the responseBuilder calls from a different thread. This
+  // happens when the use provides empty input, or the sentence and subword preprocessing deems no translatable units
+  // present. However, in this case we want an empty valid response. There's no need to do any additional processing
+  // here.
   if (segments_.size() == 0) {
-    responseBuilder_(std::move(histories_));
+    responseBuilder_(std::move(processedRequestSentences_));
+  } else {
+    counter_ = segments_.size();
+    processedRequestSentences_.resize(segments_.size());
+
+    if (cache_ != nullptr) {
+      // Iterate through segments, see if any can be prefilled from cache. If prefilled, mark the particular segments as
+      // complete (non-empty ProcessedRequestSentence). Also update accounting used elsewhere (counter_) to reflect one
+      // less segment to translate.
+      for (size_t idx = 0; idx < segments_.size(); idx++) {
+        ProcessedRequestSentence processedRequestSentence;
+        if (cache_->fetch(getSegment(idx), processedRequestSentence)) {
+          processedRequestSentences_[idx] = processedRequestSentence;
```

Comment:
Why write somewhere else then copy when you could have written directly?

Reply:
This started off as a `Ptr<...>` when it was cheap, I see the issue and the merits of the recommendation. Will fix.

Reply:
Copy forbidden, now direct move happening.